File: 3_longest-substring-without-repeating-characters/result.py
from collections import defaultdict

"""
给定一个字符串 s ，请你找出其中不含有重复字符的 最长子串 的长度。

 

示例 1:

输入: s = "abcabcbb"
输出: 3 
解释: 因为无重复字符的最长子串是 "abc"，所以其长度为 3。
示例 2:

输入: s = "bbbbb"
输出: 1
解释: 因为无重复字符的最长子串是 "b"，所以其长度为 1。
示例 3:

输入: s = "pwwkew"
输出: 3
解释: 因为无重复字符的最长子串是 "wke"，所以其长度为 3。
     请注意，你的答案必须是 子串 的长度，"pwke" 是一个子序列，不是子串。
示例 4:

输入: s = ""
输出: 0

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/longest-substring-without-repeating-characters
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        lenth = 0
        l = []
        for i in range(len(s)):
            if s[i] not in l:
                l.append(s[i])
                lenth = max(lenth, len(l))
            else:
                l = l[l.index(s[i]) + 1 :]
                l.append(s[i])
        return lenth


if __name__ == "__main__":
    assert Solution().lengthOfLongestSubstring("pwwkew") == 3

File: 3_longest-substring-without-repeating-characters/result.go
package main

func lengthOfLongestSubstring(s string) int {
	var l []string
	var lengh int
	for i := 0; i < len(s); i++ {
		var exist bool = false
		var order int = 0
		for order = 0; order < len(l); order++ {
			if l[order] == string(s[i]) {
				exist = true
				break
			}

		}
		if exist {
			if order+1 >= len(l) {
				l = []string{}
			} else {
				l = l[order+1:]
			}

			l = append(l, string(s[i]))
		} else {
			l = append(l, string(s[i]))
			if lengh < len(l) {
				lengh = len(l)
			}
		}
	}
	return lengh
}
func main() {
	print(lengthOfLongestSubstring("pwwkew"))
}
